nomad/state/state_store.go

```diff
@@ -1058,16 +1058,15 @@ func deleteNodeCSIPlugins(txn *memdb.Txn, node *structs.Node, index uint64) erro
 
 func deleteNodeFromPlugin(txn *memdb.Txn, plug *structs.CSIPlugin, node *structs.Node, index uint64) error {
 	plug.DeleteNode(node.ID)
+	return deleteFromPlugin(index, txn, plug)
+}
+
+// deleteFromPlugin updates a plugin but will delete it if the plugin is empty
+func deleteFromPlugin(index uint64, txn *memdb.Txn, plug *structs.CSIPlugin) error {
 	plug.ModifyIndex = index
 
 	if plug.IsEmpty() {
-		_, err := txn.Get("csi_volumes", "plugin_id", plug.ID)
-		if err != nil {
-			// don't delete a plugin that has associated volumes, even
-			// if there are no plugin instances running for it
-			return nil
-		}
```

Comment:
This seems like it's designed to avoid nil errors in volumes with a missing plugin, which we actually avoid from the volume read side.

Reply:
👍